scripts/man/init-checkconf.8
.TH init\-checkconf 8 2013-11-19 "Upstart"
.\"
.SH NAME
init\-checkconf \- manual page for init-checkconf
.\"
.SH SYNOPSIS
.B init\-checkconf
.RI [ OPTIONS ] " file "

.B init\-checkconf
.RI [ OPTIONS ] " " "\-f" " file "
.\"
.SH DESCRIPTION
Determine if specified Upstart
.BR init (8)
job configuration file is valid.
.P
By default, two checks are performed:
.IP \(bu 4
Ensure Upstart can parse overall file successfully
.IP \(bu 4
Ensure all script sections are parseable by shell
.P
.\"
.SH OPTIONS
.TP
.BR \-d " , " \-\-debug
Show some debug output.
.TP
.BR \-f " \fIfile\fP" " , " \-\-file=\fIfile\fP
Specify job configuration file to check.
.TP
.BR \-i " \fIpath\fP" " , " \-\-initctl\-path=\fIpath\fP
Specify path to
.BR initctl (8)
binary.
.TP
.BR \-s " , " \-\-noscript
Do not check script sections.
.TP
.BR \-x " \fIpath\fP" " , " \-\-upstart\-path=\fIpath\fP
Specify path to
.BR init (8)
binary.
.TP
.BR \-h " , " \-\-help
Display usage statement.
.\"
.SH AUTHOR
Written by James Hunt
.RB < james.hunt@ubuntu.com >
.\"
.SH LIMITATIONS
.IP \(bu 4
.B exec
stanzas containing shell meta-characters will not be checked as scripts.
.\"
.SH REPORTING BUGS
Report bugs at
.RB < https://launchpad.net/upstart/+bugs >
.\"
.SH COPYRIGHT
Copyright \(co 2011-2013 Canonical Ltd.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\"
.SH SEE ALSO
.BR sh (1)
.BR init (5)
.BR init (8)


NEWS
1.13  xxxx-xx-xx ""

	* NEWS: Begin new release.

1.12.1  2014-03-11 "Standing Room Only"

	* Fixed re-exec breakage caused by incorrect D-Bus API usage.
	* Tests updated to catch this issue.

1.12  2014-03-07 "Water Table"

	* Stop Upstart indirectly stealing console ownership (most noticeably when
	  run in a container).
	* Fixed bug that could result in incorrect output from initctl(8)
	  environment commands.
	* Fixed incorrect variable usage which could result in a crash
	  (LP: #1222705).
	* Fixed stateful re-exec issue triggered by invalid jobs
	  (LP: #1269731).
	* Stateful re-exec fix to serialise the D-Bus session bus
	  (LP: #1258098).
	* init-checkconf(8) now uses a Session Init and can also now be
	  run as root.
	* upstart-socket-bridge(8) now supports IPv6 (thanks to Kai Mast and
	  Cameron Norman).
	* Allow telinit(8) to work on systems with no D-Bus system bus.
	* Numerous test improvements.
	* Lots of man-page improvements.

1.11  2013-11-14 "Quoi de neuf?"

	* Add '--no-dbus' command-line option for the init(8) daemon (allowing
	  event propagation to the Session Init via the upstart-event-bridge(8) to
	  be disabled).
	* Serialise Session Init job environment table on re-exec (LP: #1238078).
	* upstart-{dbus,socket}-bridge(8): Don't clobber existing PATH (LP: #1234898).
	* upstart-file-bridge(8):
	  - Allow watching for directory creation (LP: #1221466).
	  - Reduce default output.
	* upstart-local-bridge(8): Handle arbitrary numbers of pairs and additional
	  checks on input validity (LP: #1234898).
	* Improved Session Init shutdown speed (LP: #1227212).
	* New configure options: '--disable-local-bridge', '--disable-socket-bridge'.
	* Added upstart-file-bridge(8) and Session Init re-exec integration tests.
	* Retain umask for Session Init (LP: #1240686).
	* Make Session Init connect to D-Bus Session Bus on request
	  (LP: #1203595, #1235649)
	* Improved build-time ABI compliance checking.
	* Doc updates and lots of test refinements.

1.10  2013-08-23 "Swim to the buoy"

	* upstart-local-bridge: New bridge for starting jobs on local
	  socket connections.
	* upstart-dconf-bridge: New bridge for Session Inits.
	* upstart-dbus-bridge: New '--bus-name' option to allow bus name
	  variable to be included in dbus-event(7).
	* New "reload signal" stanza to allow jobs to specify a custom signal
	  that will be sent to the main process (rather than the default
	  SIGHUP).
	* Inclusion of Session Init sample jobs.
	* Re-exec fixes for handling chroot sessions.
	* Shutdown fix for Session Inits.
	* New python3 module and accompanying integration test suite for
	  testing Upstart running as PID 1 and as a Session Init
	  (priveleged and non-privileged).

1.9.1  2013-07-04 "The all-inclusive deal"

	* Fix respawn behaviour regression (introduced in version 1.8) affecting
	  jobs which also specify shell meta-characters in the exec stanza
	  (LP: #1181789).
	* Fix common bug which could cause a bridge to crash (LP: #1197225).
	* Various fixes for libupstart.
	* initctl2dot: Layout improvements.

1.9  2013-06-28 "This village isn't even on the map"

	* Apparmor support added by means of two new stanzas: 'apparmor load'
	  and 'apparmor switch' (thanks Marc Deslauriers).
	* Stateful re-exec: now serialises all objects, not just the minimum set.
	* Allow inherited environment variables to be unset for Session Inits.
	* Ability to specify multiple configuration directories when running as
	  a Session Init.
	* libupstart: Client library for communicating with Upstart.
	* upstart-dbus-bridge: New bridge to allow jobs to react to D-Bus signals.
	* The usual round of fixes and doc improvements.

1.8  2013-03-22 "Scone + Radish = Gherkin?"

	* upstart-file-bridge: New bridge to allow jobs to react to
	  file, directory and file glob events (create/modify/delete).
	* upstart-monitor: Simple cli/gui tool that shows event flows.

1.7  2013-03-04 "We've all got one now"

	* New initctl commands: set-env, unset-env, get-env, list-env,
	  reset-env, list-sessions (all except last with corresponding
	  D-Bus methods).
	* New D-Bus-only signals EventEmitted, Restarted, and EndSession method.
	* Ability to run with PID >1 to allow Upstart to manage a user session.
	  Running Upstart as a 'Session Init' in this way provides features
	  above and beyond those provided by the original User Jobs such that
	  the User Job facility has been removed entirely: to migrate from
	  a system using User Jobs, simply ensure the user session is started with
	  'init --user'.
	* New upstart-event-bridge bridge which proxies system-level events down
	  to Session Inits, allowing users jobs to react to udev events.
	* Ability to read job configuration and override files from multiple
	  freedesktop-compliant locations (Session Init only).
	* Ability to shutdown both via a system shutdown request and via a user
	  logout request (Session Init only).

1.6.1  2012-12-07 "It glows in the dark"

	* Improved re-exec performance.
	* Minor logger fixes for unflushed data.
	* Handle re-exec scenario when requested from within a chroot.
	* Minor serialisation data format change to handle chroots and
	  user sessions.
	* Added extra re-exec tests including explicit upgrade tests reading
	  from pre-prepared JSON data files.
	* Make jobs running within a chroot log their output within the chroot.
	* Added "Restart" D-Bus method.
	* Changed 'telinit u' to use "Restart" D-Bus method rather than
	  sending SIGTERM to play nicely when busybox(1) is init.
	* Added "GetState" D-Bus method allowing current serialised internal state
	  to be queried.

1.6  2012-11-15 "The turbine, the fish and the glass partition"

	* 'telinit u' now performs a stateful re-exec, allowing Upstart
	  to continue to supervise jobs after an upgrade of either
	  itself, or any of its dependent libraries. Adds dependency
	  on libjson (JSON-C).
	* Added initramfs-less boot support.
	* Allow tests to work in sbuild environment.
	* Improved error handling.

scripts/init-checkconf.sh
#!/bin/bash
#---------------------------------------------------------------------
# Script to determine if specified config file is valid or not.
# By default, two checks are performed:
#
#   - Ensure Upstart can parse overall file successfully
#   - Ensure all script sections are parseable by shell
#
#---------------------------------------------------------------------
#
# Copyright (C) 2011-2013 Canonical Ltd.
#
# Author: James Hunt <james.hunt@canonical.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#---------------------------------------------------------------------

script_name=${0##*/}
confdir=$(mktemp -d /tmp/${script_name}.XXXXXXXXXX)
xdg_runtime_dir=$(mktemp -d /tmp/${script_name}.XXXXXXXXXX)
upstart_path=/sbin/init
initctl_path=/sbin/initctl
debug_enabled=n
file_valid=n
running=n
set_session=n
check_scripts=y

cleanup()
{
    # Restore
    [ -n "$saved_xdg_runtime_dir" ] && \
        debug "Restoring XDG_RUNTIME_DIR to '$saved_xdg_runtime_dir'"
    export XDG_RUNTIME_DIR="$saved_xdg_runtime_dir"

    [ -n "$saved_upstart_session" ] && \
        debug "Restoring UPSTART_SESSION to '$saved_upstart_session'"
    export UPSTART_SESSION="$saved_upstart_session"
  
    if [ ! -z "$upstart_pid" ]
    then
      debug "Stopping secondary Upstart (running with PID $upstart_pid)"
      kill -9 "$upstart_pid" >/dev/null 2>&1
    fi
  
    [ -d "$confdir" ] && rm -rf "$confdir"
    dir="$xdg_runtime_dir/upstart/sessions"
    [ -d "$dir" ] && rm -rf "$xdg_runtime_dir"
    [ "$file_valid" = y ] && exit 0
  
    exit 1
}

usage()
{
cat <<EOT
Description: Determine if specified Upstart (init(8)) job configuration
             file is valid.

Usage: $script_name [options] -f <conf_file>
       $script_name [options]    <conf_file>

Options:

 -d, --debug           : Show some debug output.
 -f <file>,            : Job configuration file to check.
 --file=<file>           (no default).
 -i <path>,            : Specify path to initctl binary
 --initctl-path=<path>   (default=$initctl_path).
 -s, --noscript        : Do not check script sections.
 -x <path>             : Specify path to init daemon binary
 --upstart-path=<path>   (default=$upstart_path).
 -h, --help            : Show this help.

EOT
}

debug()
{
    msg="$*"
    [ $debug_enabled = y ] && echo "DEBUG: $msg"
}

error()
{
    msg="$*"
    printf "ERROR: %s\n" "$msg" >&2
}

die()
{
    error "$*"
    exit 1
}

# Return 0 if Upstart is running, else 1
upstart_running()
{
    initctl --user version >/dev/null 2>&1
}

trap cleanup EXIT INT TERM

args=$(getopt \
    -n "$script_name" \
    -a \
    --options="df:hi:sx:" \
    --longoptions="debug file: help initctl-path: noscript upstart-path:" \
    -- "$@")

eval set -- "$args"
[ $? -ne 0 ] && { usage; exit 1; }
[ $# -eq 0 ] && { usage; exit 0; }

while [ $# -gt 0 ]
do
    case "$1" in
      -d|--debug)
        debug_enabled=y
        ;;

      -f|--file)
        file="$2"
        shift
        ;;

      -h|--help)
        usage
        exit 0
        ;;

      -i|--initctl-path)
        initctl_path="$2"
        shift
        ;;

      -s|--noscript)
        check_scripts=n
        ;;

      -x|--upstart-path)
        upstart_path="$2"
        shift
        ;;

      --)
        shift
        break
        ;;
    esac
    shift
done

[ -z "$file" ] && file="$1"
[ -z "$file" ] && die "Must specify configuration file"
[ ! -f "$file" ] && die "File $file does not exist"

debug "upstart_path=$upstart_path"
debug "initctl_path=$initctl_path"

for cmd in "$upstart_path" "$initctl_path"
do
    [ -f "$cmd" ] || die "Path $cmd does not exist"
    [ -x "$cmd" ] || die "File $cmd not executable"
    "$cmd" --help | grep -q -- --user || die "version of $cmd too old"
done

export saved_xdg_runtime_dir="$XDG_RUNTIME_DIR"
debug "Setting XDG_RUNTIME_DIR='$xdg_runtime_dir'"
export XDG_RUNTIME_DIR="$xdg_runtime_dir"

export saved_upstart_session="$UPSTART_SESSION"
[ -n "$UPSTART_SESSION" ] \
    && debug "Unsetting UPSTART_SESSION ($UPSTART_SESSION)" \
    && unset UPSTART_SESSION

# this is the only safe way to run another instance of Upstart
"$upstart_path" --help|grep -q -- --no-startup-event || die "$upstart_path too old"

debug "confdir=$confdir"
debug "file=$file"

filename=$(basename $file)

echo "$filename" | egrep -q '\.conf$' || die "File must end in .conf"

job="${filename%.conf}"

cp "$file" "$confdir" || die "Failed to copy file $file to $confdir"
debug "job=$job"

upstart_out="$(mktemp --tmpdir "${script_name}-upstart-output.XXXXXXXXXX")"
debug "upstart_out=$upstart_out"

upstart_cmd=$(printf \
    "%s --user --no-dbus --no-sessions --no-startup-event --verbose --confdir %s" \
        "$upstart_path" \
        "$confdir")
debug "upstart_cmd=$upstart_cmd"

nohup $upstart_cmd >"$upstart_out" 2>&1 &
upstart_pid=$!
debug "Upstart pid=$upstart_pid"

# Stop the shell outputting a message when Upstart is killed.
# We handle this ourselves in cleanup().
disown 

# wait for Upstart to initialise
for i in $(seq 1 5)
do
    sessions=$("$initctl_path" list-sessions)

    if [ "$set_session" = n ] && [ -n "$sessions" ]
    then
        count=$(echo "$sessions"|wc -l)
        [ "$count" -gt 1 ] && die "Got unexpected session count: $count"
        session=$(echo "$sessions"|awk '{print $2}')
        debug "Joining Upstart session '$session'"
        export UPSTART_SESSION="$session"
	set_session=y
    fi

    debug "Waiting for Upstart to initialise (attempt $i)"

    upstart_running
    if [ $? -eq 0 ]
    then
        running=y
        break
    fi

    sleep 1
done

[ $running = n ] && die "Failed to ask Upstart to check conf file"

debug "Secondary Upstart ($upstart_cmd) running with PID $upstart_pid"

if [ "$check_scripts" = y ]
then
    for section in pre-start post-start script pre-stop post-stop
    do
        if egrep -q "\<${section}\>" "$file"
        then
            cmd='sed -n "/^ *${section}/,/^ *end script/p" $file | /bin/sh -n 2>&1'
            errors=$(eval "$cmd")
            [ $? -ne 0 ] && \
                die "$(printf "File $file: shell syntax invalid in $section section:\n${errors}")"
        fi
    done
fi

"$initctl_path" --user list|grep -q "^${job}"
if [ $? -eq 0 ]
then
    file_valid=y
    echo "File $file: syntax ok"
    exit 0
fi

errors=$(grep "$job" "$upstart_out"|sed "s,${confdir}/,,g")
die "$(printf "File $file: syntax invalid:\n${errors}")"
